src/libxtp/gwbse/sigma_cda.cc

```diff
@@ -0,0 +1,120 @@
+/*
+ *            Copyright 2009-2019 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <votca/tools/constants.h>
+#include <votca/xtp/gw.h>
+#include <votca/xtp/sigma_cda.h>
+
+namespace votca {
+namespace xtp {
+
+void Sigma_CDA::PrepareScreening() {
+  GaussianQuadrature::options opt;
+  opt.homo = _opt.homo;
+  opt.order = _opt.order;
+  opt.qptotal = _qptotal;
+  opt.qpmin = _opt.qpmin;
+  opt.rpamin = _opt.rpamin;
+  opt.alpha = _opt.alpha;
+  opt.quadrature_scheme = _opt.quadrature_scheme;
+  _gq.configure(opt, _rpa);
+}
+
+// This function is used in the calculation of the residues
+// This calculates eps^-1 (inverse of the dielectric function) for complex
+// frequencies of the kind omega = delta + i*eta
+double Sigma_CDA::CalcDiagContribution(Eigen::RowVectorXd Imx_row, double delta,
+                                      double eta) const {
+  Eigen::MatrixXcd DielMxInv =
+      _rpa.calculate_epsilon_complex(delta, eta).inverse();
+  DielMxInv.diagonal().array() -= 1.0;
+  std::complex<double> lambda =
+      ((Imx_row * DielMxInv).cwiseProduct(Imx_row)).sum();
+  return lambda.real();
+}
+
+double Sigma_CDA::CalcResiduePrefactor(double e_f, double e_m,
+                                      double frequency) const {
+  double factor = 0.0;
+
+  if (e_f < e_m && e_m < frequency) {
+    factor = 1.0;
+  } else if (e_f > e_m && e_m > frequency) {
+    factor = -1.0;
+  } else if (e_m - frequency == 0.0 && e_f > e_m) {
+    factor = -0.5;
+  } else if (e_m - frequency == 0.0 && e_f < e_m) {
+    factor = 0.5;
```

Comment:
comparing a double to an exact value is dangerous, are you sure this is you want and not a `std::abs(e_m-frequency)<tolerance` ? This will give different results depending on the architecture

Reply:
I will try this way